### ISSUE TITLE
Cancelling a TextDecoderStream that failed to read invalid data raises TypeError

### ISSUE THREAD

**Issue by effigies** [denoland/deno#24872]
Version: Deno 1.45.5

I'm working on a tool where we want to gracefully report an error that a file is not UTF-8. I can catch the `TypeError` from attempting to read the data, but when attempting to cancel the stream to avoid a resource leak, I get another `TypeError`. Catching that only continues to leak.

```typescript
import { assert } from "https://deno.land/std/testing/asserts.ts";

Deno.test("Reproduce resource leak", async (t) => {
  const encoding = Deno.env.get("ENCODING") || "utf-8";
  
  // Create a stream that will yield UTF-16 encoded bytes
  const utf16str = "\uFEFFAny text";
  const buffer = new ArrayBuffer(utf16str.length * 2);
  const view = new Uint16Array(buffer);
  for (let i = 0, strLen = utf16str.length; i < strLen; i++) {
    view[i] = utf16str.charCodeAt(i);
  }
  const stream = new ReadableStream({     
    start(controller) {     
      controller.enqueue(new Uint8Array(buffer));     
      controller.close();     
    },     
  });     
  
  // Pipe the stream through a UTF-8 decoder that will throw an error
  const decoder = new TextDecoderStream(encoding, { fatal: true });    
  const reader = stream.pipeThrough(decoder).getReader();     
  let errored = false;     
  
  // Catch the error thrown by the decoder
  try {                                                           
    const { done, value } = await reader.read();
    assert(value === "Any text");
  } catch (e) {     
    errored = true;
  }               
  assert(errored === (encoding === "utf-8"));
  
  console.log("Closing reader") 
  try {                                                                                                        
    await reader.cancel();
  } catch (e) {                                                                                                
    console.log("Error cancelling reader:\n", e);
  }  
});
```

```console
❯ deno test --allow-env .
Check file:///home/chris/Projects/bugs/deno-textdecoder/reproduce.test.ts
running 1 test from ./reproduce.test.ts
Reproduce resource leak ...
------- output -------
Closing reader
Error cancelling reader:
 TypeError: The encoded data is not valid.
    at TextDecoder.decode (ext:deno_web/08_text_encoding.js:183:14)
    at Object.transform (ext:deno_web/08_text_encoding.js:308:41)
    at Module.invokeCallbackFunction (ext:deno_webidl/00_webidl.js:981:16)
    at TransformStreamDefaultController.transformAlgorithm (ext:deno_web/06_streams.js:3788:14)
    at transformStreamDefaultControllerPerformTransform (ext:deno_web/06_streams.js:4057:59)
    at transformStreamDefaultSinkWriteAlgorithm (ext:deno_web/06_streams.js:4160:10)
    at WritableStreamDefaultController.writeAlgorithm (ext:deno_web/06_streams.js:630:12)
    at writableStreamDefaultControllerProcessWrite (ext:deno_web/06_streams.js:4484:55)
    at writableStreamDefaultControllerAdvanceQueueIfNeeded (ext:deno_web/06_streams.js:4387:5)
    at ext:deno_web/06_streams.js:3870:5
----- output end -----
Reproduce resource leak ... FAILED (3ms)

 ERRORS 

Reproduce resource leak => ./reproduce.test.ts:3:6
error: Leaks detected:
  - A text decoder was created during the test, but not finished during the test. Close the text decoder by calling `textDecoder.decode('')` or `await textDecoderStream.readable.cancel()`.

 FAILURES 

Reproduce resource leak => ./reproduce.test.ts:3:6

FAILED | 0 passed | 1 failed (5ms)

error: Test failed
```

To verify that the code generally works if we're expecting UTF-16 data, run with `ENCODING=utf-16`.

It's possible I'm doing something wrong in general, but I think this is a bug.

**Comment by BlackAsLight** [comment]
> To verify that the code generally works if we're expecting UTF-16 data, run with ENCODING=utf-16.

How can you be sure of that with that second assert. What is the point of it?

**Comment by effigies** [comment]
You mean `assert(errored === (encoding === "utf-8"));`?

That confirms that we only entered the `catch` block and set `errored = True` if the encoding we attempted to use was `utf-8`. Which means that if `utf-16` is true, then we also hit the `assert(value === "Any text");`

**Comment by effigies** [comment]
This might be the same as https://github.com/denoland/deno/issues/19946, but I'm not sure about that.

I will note that the hack in https://github.com/denoland/deno/issues/13142#issuecomment-1169072506 works, but presumably only because the files can be slurped in a single chunk. If you update it to use `stream: true`, then it also leaks:

```typescript
  const MyTextDecoderStream = () => {                                                
    const textDecoder = new TextDecoder();                                           
    return new TransformStream({                                                     
      transform(chunk : Uint8Array, controller: TransformStreamDefaultController) {  
        controller.enqueue(textDecoder.decode(chunk, { stream: true }));
      },                                                                                                                          
      flush(controller: TransformStreamDefaultController) {                          
        controller.enqueue(textDecoder.decode());                                    
      }                                                                              
    });                                                                              
  }; 
```

**Comment by crowlKats** [comment]
@effigies The stream behaviour itself is correct, and chrome behaves the same as well. 

As a side-note, the leak is annoying, and #24878 would fix that.

**Comment by lucacasonato** [comment]
This is fixed in Deno 1.46.0

**Comment by effigies** [comment]
Can confirm that my reproduction no longer errors. Thanks!
